src/db_cache/foyer.rs

```diff
@@ -0,0 +1,107 @@
+//! # Foyer Cache
+//!
+//! This module provides an implementation of an in-memory cache using the Foyer library.
+//! The cache is designed to store and retrieve cached blocks associated with SSTable IDs.
+//! It supports configurable capacity and block size.
+//!
+//! ## Features
+//!
+//! - **Asynchronous Operations**: Utilizes Foyer's `Cache` to perform cache operations asynchronously.
+//! - **Custom Weigher**: Implements a custom weigher to account for the size of cached blocks.
+//! - **Flexible Configuration**: Allows customization of cache parameters such as maximum capacity and block size.
+//!
+//! ## Examples
+//!
+//! ```rust,no_run
+//! use slatedb::db::Db;
+//! use slatedb::db_cache::foyer::{FoyerCache, FoyerCacheOptions};
+//! use object_store::local::LocalFileSystem;
+//! use std::sync::Arc;
+//!
+//! #[::tokio::main]
+//! async fn main() {
+//!     let object_store = Arc::new(LocalFileSystem::new());
+//!     let cache = Arc::new(FoyerCache::new());
+//! let db = Db::open_with_cache("path/to/db".into(), object_store, cache).await;
+//! }
+//! ```
+//!
+use crate::db_cache::{
+    CachedEntry, DbCache, SsTableId, DEFAULT_CACHED_BLOCK_SIZE, DEFAULT_MAX_CAPACITY,
+};
+use async_trait::async_trait;
+
+/// The options for the Foyer cache.
+#[derive(Clone, Copy, Debug)]
+pub struct FoyerCacheOptions {
+    pub max_capacity: u64,
+    pub cached_block_size: u32,
+}
+
+impl Default for FoyerCacheOptions {
+    fn default() -> Self {
+        Self {
+            max_capacity: DEFAULT_MAX_CAPACITY,
+            cached_block_size: DEFAULT_CACHED_BLOCK_SIZE,
+        }
+    }
+}
+
+/// A cache implementation using the Foyer library.
+///
+/// This struct wraps a Foyer cache, providing an in-memory caching solution
+/// for storing and retrieving cached blocks associated with SSTable IDs.
+///
+/// # Fields
+///
+/// * `inner` - The underlying Foyer cache instance, which maps `(SsTableId, u64)`
+///   keys to `CachedBlock` values.
+///
+/// # Notes
+///
+/// The cache is configured based on the provided `InMemoryCacheOptions`,
+/// including settings for capacity, time-to-live (TTL), and time-to-idle (TTI).
+/// It uses a custom weigher to account for the size of cached blocks.
+pub struct FoyerCache {
+    inner: foyer::Cache<(SsTableId, u64), CachedEntry>,
+}
+
+impl FoyerCache {
+    pub fn new() -> Self {
+        Self::new_with_opts(FoyerCacheOptions::default())
+    }
+
+    pub fn new_with_opts(options: FoyerCacheOptions) -> Self {
+        let builder = foyer::CacheBuilder::new(options.max_capacity as _)
+            .with_weighter(move |_, _| options.cached_block_size as _);
```

Comment:
I think this isn't changed functionally in this patch, but what does this mean? that we assign a static size to all cached entries? (cc @MrCroxx). If true seems like something we should fix - seems like it would be straightforward to extend CachedEntry with a `size` method and just use that in the weigher. Definitely not in this PR - just making a note of it. If that makes sense @MrCroxx I can file an issue.

Reply:
Good catch. Yea, I think this is wrong. It should use the size of the block/index/filter. It looks like the Moka cache also has it wrong: https://github.com/slatedb/slatedb/pull/231/files#diff-a371eaafd17b8543224b3380a8d29dfd9fd635760901043fb1c2296b1490fbd7R85

Reply:
filed https://github.com/slatedb/slatedb/issues/241